iceoryx_posh/include/iceoryx_posh/experimental/internal/popo/typed_publisher.inl

```diff
@@ -0,0 +1,122 @@
+// Copyright (c) 2020 by Robert Bosch GmbH. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef IOX_EXPERIMENTAL_POSH_POPO_TYPED_PUBLISHER_INL
+#define IOX_EXPERIMENTAL_POSH_POPO_TYPED_PUBLISHER_INL
+
+#include <cstdint>
+
+namespace iox
+{
+namespace popo
+{
+template <typename T, typename base_publisher_t>
+TypedPublisher<T, base_publisher_t>::TypedPublisher(const capro::ServiceDescription& service)
+    : base_publisher_t(service)
+{
+}
+
+template <typename T, typename base_publisher_t>
+inline uid_t TypedPublisher<T, base_publisher_t>::getUid() const noexcept
+{
+    return base_publisher_t::getUid();
+}
+
+template <typename T, typename base_publisher_t>
+inline cxx::expected<Sample<T>, AllocationError> TypedPublisher<T, base_publisher_t>::loan() noexcept
+{
+    return base_publisher_t::loan(sizeof(T));
+}
+
+template <typename T, typename base_publisher_t>
+inline void TypedPublisher<T, base_publisher_t>::publish(Sample<T> sample) noexcept
+{
+    return base_publisher_t::publish(sample);
+}
+
+template <typename T, typename base_publisher_t>
+template <typename Callable, typename... ArgTypes>
+inline cxx::expected<AllocationError> TypedPublisher<T, base_publisher_t>::publishResultOf(Callable c,
+                                                                                           ArgTypes... args) noexcept
+{
+    static_assert(cxx::is_callable<Callable, T*, ArgTypes...>::value,
+                  "TypedPublisher<T>::publishResultOf expects a valid callable as the first argument");
+    static_assert(cxx::has_signature<Callable, void(T*, ArgTypes...)>::value,
+                  "callable provided to TypedPublisher<T>::publishResultOf must have signature void(T*, ArgsTypes...)");
+
+    auto result = loan();
```

Comment:
I think you can always return the result here. With this in mind you could rewrite the code to something like this:
```cpp
return loan().and_then([&](auto & sample){
        c(sample.get(), std::forward<ArgTypes>(args)...);
        publish(std::move(sample));
      });
```

The idea was the expected's are convertable to each other since you often have exactly this use case. You return a value, do stuff with it and if it is returning an error your forward the error. 

If it is not yet implemented we have to do this.


If it works you can do the same this for `publishCopyOf`

Reply:
Nice, I didn't realise and_then() returns a `success`.